airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py

```diff
@@ -2,36 +2,142 @@
+          - The future will add the partitions to process on a work queue
+        2. Continuously poll work from the work queue until all partitions are generated and processed
+          - If the next work item is a partition, submit a future to process it.
+            - The future will add the records to emit on the work queue
+            - Add the partitions to the partitions_to_done dict so we know it needs to complete for the sync to succeed
+          - If the next work item is a record, yield the record
+          - If the next work item is PARTITIONS_GENERATED_SENTINEL, all the partitions were generated
+          - If the next work item is a PartitionCompleteSentinel, a partition is done processing
+            - Update the value in partitions_to_done to True so we know the partition is completed
+        """
+        self.logger.debug(f"Processing stream slices for {self.name} (sync_mode: full_refresh)")
+        futures = []
+        queue: Queue[QueueItem] = Queue()
+        partition_generator = PartitionEnqueuer(queue, PARTITIONS_GENERATED_SENTINEL)
+        partition_reader = PartitionReader(queue)
+
+        # Submit partition generation tasks
+        futures.append(
+            self._threadpool.submit(partition_generator.generate_partitions, self._stream_partition_generator, SyncMode.full_refresh)
+        )
+
+        # True -> partition is done
+        # False -> partition is not done
+        partitions_to_done: Dict[Partition, bool] = {}
+
+        finished_partitions = False
+        while record_or_partition := queue.get(block=True, timeout=self._timeout_seconds):
+            if record_or_partition == PARTITIONS_GENERATED_SENTINEL:
+                # All partitions were generated
+                finished_partitions = True
+            elif isinstance(record_or_partition, PartitionCompleteSentinel):
+                # All records for a partition were generated
+                if record_or_partition.partition not in partitions_to_done:
+                    raise RuntimeError(
+                        f"Received sentinel for partition {record_or_partition.partition} that was not in partitions. This is indicative of a bug in the CDK. Please contact support.partitions:\n{partitions_to_done}"
+                    )
+                partitions_to_done[record_or_partition.partition] = True
+            elif isinstance(record_or_partition, Record):
+                # Emit records
+                yield record_or_partition
+            elif isinstance(record_or_partition, Partition):
+                # A new partition was generated and must be processed
+                partitions_to_done[record_or_partition] = False
+                if self._slice_logger.should_log_slice_message(self.logger):
+                    self._message_repository.emit_message(self._slice_logger.create_slice_log_message(record_or_partition.to_slice()))
```

Comment:
same conceptually the same as [what we do for regular stream](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L247-L248), but we emit the message on a message repository instead of yielding them. [They will be consumed by the abstract source](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L257-L258)